design/use-statement.md

```diff
@@ -0,0 +1,162 @@
+# The `use` statement
+
+There are currently 9 different [tool or installer](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/go-tool?view=vsts) tasks.
+Conceptually they do roughly the same thing (in the user's mind):
+- Get a known version of some tool, maybe from the internet
+- Set up the CI environment so that when I invoke `tool`, I get that version.
```

Comment:
FYI we faced a similar problem with the Rush tool.  We solved it using a custom script "install-run-rush.js" which is documented here:
[https://rushjs.io/pages/maintainer/enabling_ci_builds/](https://rushjs.io/pages/maintainer/enabling_ci_builds/)

Some problems we had to solve:
- Determine the version of the tool to install based on the repo's rush.json config file, which is what Rush uses itself
- Cache the installed tool across multiple builds, with the observation that a cached installation cannot be reused if it was installed using a different version of the NodeJS engine
- Support installation from a private NPM Feed (or not), according to the repo's .npmrc file

I wonder if it would make sense to provide built-in support for installing Rush on a VSTS CI machine

Reply:
Nice, thanks. We want to offer an extensibility model so that any tool can be delivered this way.